src/api/app/jobs/cleanup_events.rb

```diff
@@ -1,7 +1,23 @@
 class CleanupEvents < ApplicationJob
   def perform
     Event::Base.transaction do
-      Event::Base.where(project_logged: true, queued: true, undone_jobs: 0).lock(true).delete_all
+      events_to_be_deleted.lock(true).delete_all
     end
   end
+
+  private
+
+  def events_to_be_deleted
+    # delete_all will not work with this query so we need to find again by id
+    events_to_be_deleted_array =
+      Event::Base.left_outer_joins(:digest_emails)
+        .select('events.*, MIN(digest_emails.email_sent) AS has_all_digest_emails_sent')
+        .where(project_logged: true, queued: true, undone_jobs: 0)
+        .group('events.id')
+        .having('has_all_digest_emails_sent = 1 OR has_all_digest_emails_sent IS NULL')
+
+    event_ids = events_to_be_deleted_array.map(&:id)
```

Comment:
@coolo, @hennevogel raised the point that `CleanupEvents` is only run [once per day](https://github.com/evanrolfe/open-build-service/blob/feature%2Fdaily_digest_emails/src/api/config/clock.rb#L60-L62), so we always "pile up a day worth of events" anyway before that job runs. I don't think that this is a problem?